### PR TITLE
560:: feat: integration test with check pin method which need logic improve…

### PR DESCRIPTION
**Pull request by HaseebUllahAbbasi**
the checkPin method is added in the user.ts 

- [x] if the pin is incorrect then user should get error
- [x] if the pin is correct then it should return true

### CODE REVIEW

File: src/user.ts [hunk 1]
```diff
@@ -158,6 +158,32 @@ export class User {
...
+        });
+
+        // console.log('Pin Key', pinKey?.toString());
+        // console.log('Pin Key data', pinKey?.data);
+
+        console.log('pin', pin);
+        const accountData = await User.getAccountInfo(idData.account_name);
+        // get pin Pin now and match with given Pin key
+
+        const onchainKey = accountData.getPermission('owner').required_auth.keys[0].key;
```
Comment (theblockstalk) [inline src/user.ts:177]: should check pin permission, not owner

File: src/user.ts [hunk 2]
```diff
@@ -158,6 +158,32 @@ export class User {
...
+    async checkPin(pin: string): Promise<boolean> {
+        const username = await this.getAccountName();
+
+        const idData = await idContract.getPerson(username);
+
+        const pinKey = await this.keyManager.getKey({
+            level: KeyManagerLevel.PIN,
+        });
+
+        // console.log('Pin Key', pinKey?.toString());
```
Comment (theblockstalk) [inline src/user.ts:170]: remove unused comments

**Review by theblockstalk — CHANGES_REQUESTED** [review]
missing integration tests

File: src/user.ts (hunk 1 above)
Comment (HaseebUllahAbbasi) [reply src/user.ts:177]: changed to pin

File: src/user.ts (hunk 2 above)
Comment (HaseebUllahAbbasi) [reply src/user.ts:170]: removed

File: test-integration/user.test.ts [hunk 3]
```diff
@@ -321,4 +321,39 @@ describe('User class', () => {
...
     );
+    test(
+        'CheckPin() returns true when pin matches',
+        catchAndPrintErrors(async () => {
+            const { user, password } = await createRandomID();
+
+            await user.login(await user.getUsername(), password);
+
+            await expect(user.checkPassword(password)).resolves.toBeTruthy();
+            await expect(user.checkPassword(password)).resolves.toBe(true);
```
Comment (theblockstalk) [inline test-integration/user.test.ts:332]: no need for L331 if you have L332

File: src/user.ts [hunk 4]
```diff
@@ -158,6 +158,32 @@ export class User {
         });
     }
 
+    async checkPin(pin: string): Promise<boolean> {
+        const username = await this.getAccountName();
```
Comment (theblockstalk) [inline src/user.ts:162]: don't need to check the on-chain key, only the one in keyStorage

**Review by theblockstalk — COMMENTED** [review]
we discussed yesterday that you would modify KeyManager.getKey() to have optional input `challenge` in the GetKeyOptions
challenge is compared against the salted hashed challenge

this then needs to be updated in all the implementations of the KeyManager, and their consumers

**Review by theblockstalk — CHANGES_REQUESTED** [review]
are there any instances of keyManager in Tonomy ID or App Websites, or where these keyManager instances are consumed that need updating?

File: test/services/jskeymanager.ts [hunk 5]
```diff
@@ -132,6 +132,13 @@ export class JsKeyManager implements KeyManager {
 
             const keyStore = this.keyStorage[options.level];
 
+            if (options?.challenge) {
```
[the remark this answers is not in the document]
Comment (HaseebUllahAbbasi) [reply test/services/jskeymanager.ts:135]: sure Adding

**Comment by theblockstalk** [comment]
these two lines need updating to send in the challenge
https://github.com/Tonomy-Foundation/Tonomy-ID-SDK/blob/development/src/user.ts#L185
https://github.com/Tonomy-Foundation/Tonomy-ID-SDK/blob/development/src/user.ts#L235
https://github.com/Tonomy-Foundation/Tonomy-ID-SDK/blob/development/src/user.ts#L270
https://github.com/Tonomy-Foundation/Tonomy-ID-SDK/blob/development/src/user.ts#L348

because of the above, the createPerson(), updateKeys(), checkPassword(), login(), checkKeysStillValid() functions would need to be updated to have the password/pin challenge sent in - in the tests as well as consumers of these function in TID and App Websites.

we should discuss this approach more.

File: test/services/jskeymanager.ts [hunk 6]
```diff
@@ -132,6 +132,15 @@ export class JsKeyManager implements KeyManager {
 
             const keyStore = this.keyStorage[options.level];
 
+            if (options.level === KeyManagerLevel.PIN) {
```
Comment (theblockstalk) [inline test/services/jskeymanager.ts:135]: remove this challenge guarde completly

Comment (theblockstalk) [reply test/services/jskeymanager.ts:135]: https://github.com/Tonomy-Foundation/Tonomy-ID/issues/396

Comment (theblockstalk) [reply test/services/jskeymanager.ts:135]: ok, as discussed. we will create a new function on the KeyManager
`checkKey(keyManagerLevel: KeyManagerLevel, challenge): Promise<boolean>`
which gets implemented in the instances (tests, TID and App Websites) to check if the challenge is correct

Comment (theblockstalk) [reply test/services/jskeymanager.ts:135]: this is used in checkPin()

File: src/services/keymanager.ts [hunk 7]
```diff
@@ -66,6 +66,7 @@ type SignDataOptions = {
  */
 type GetKeyOptions = {
     level: KeyManagerLevel;
+    challenge?: string;
```
Comment (theblockstalk) [inline src/services/keymanager.ts:69]: get key doesnt require challenge, check key does

```ts
type GetKeyOptions = {
    level: KeyManagerLevel;
};
type CheckKeyOptions = {
    level: KeyManagerLevel;
    challenge: string;
};
```

File: src/services/keymanager.ts [hunk 8]
```diff
@@ -121,6 +122,15 @@ interface KeyManager {
         password: string,
         salt?: Checksum256
     ): Promise<{ privateKey: PrivateKey; salt: Checksum256 }>;
+
+    /**
+     * checks the key against the provided challenge with hashedSaltedChallenge
+     * @param Options for checking key with level, and challenge
+     * @returns Boolean if matched
+     *
+     * @throws if challenge is not provided or not matched
```
Comment (theblockstalk) [inline src/services/keymanager.ts:131]: @throws if challenge is not provided

File: src/services/keymanager.ts [hunk 9]
```diff
@@ -121,6 +122,15 @@ interface KeyManager {
         password: string,
         salt?: Checksum256
     ): Promise<{ privateKey: PrivateKey; salt: Checksum256 }>;
+
+    /**
+     * checks the key against the provided challenge with hashedSaltedChallenge
+     * @param Options for checking key with level, and challenge
```
Comment (theblockstalk) [inline src/services/keymanager.ts:128]: show type and variable name like:
@param options {GetKeyOption} - 

also add a typedef to GetKeyOption

https://devhints.io/jsdoc

File: test/services/jskeymanager.ts [hunk 10]
```diff
@@ -135,6 +136,20 @@ export class JsKeyManager implements KeyManager {
...
+        if (!(options.level in this.keyStorage)) throw new Error('No key for this level');
+
+        const keyStore = this.keyStorage[options.level];
+
+        if (options.level === KeyManagerLevel.PIN) {
+            if (options?.challenge) {
+                const hashedSaltedChallenge = sha256(options?.challenge + keyStore.salt);
+
+                if (hashedSaltedChallenge === keyStore.hashedSaltedChallenge) return true;
+                else throw throwError('Pin is incorrect', SdkErrors.PinInValid);
```
Comment (theblockstalk) [inline test/services/jskeymanager.ts:149]: this should return false

**Review by theblockstalk — APPROVED** [review]

**Pull request merged**